### ISSUE TITLE
[Q&A] runner multi process 구조  + 50:50 분산 처리에 대하여

### ISSUE THREAD

**Issue by KimSoungRyoul** [KimSoungRyoul/PyConKR2023-ModelServing-BentoML#3]
`workers_per_resource` 옵션을 사용하는 Runner process 갯수 조절가능 
https://docs.bentoml.com/en/latest/guides/scheduling.html <-- more detail
~~~yaml
runners:
  resources:
    cpu: 4
  workers_per_resource: 2 # more than one
~~~

이 기능은 2023-08 기준 나온지 몇달 안됨

Runner를 직접 오버라이딩해서 `support_multi_thread` 옵션을 끄고 재구현 하는 방식은 부하 분산에 문제가 있다고함 

---

### `workers_per_resource` 옵션 없이 코드로 수동 분산할수 있는 예제 코드 
동일 모델로 Runner를 2개 생성한다음 (= process 2개를 생성한다음)

~~~Python
mnist_runner1 = bentoml.pytorch.get("pytorch_mnist:latest").to_runner(name="pytorch_mnist1") # spawned proccess 1
mnist_runner2 = bentoml.pytorch.get("pytorch_mnist:latest").to_runner(name="pytorch_mnist2") # spawned proccess 2

svc = bentoml.Service(name="pytorch_mnist_demo", runners=[mnist_runner1, mnist_runner2])


def to_numpy(tensor):
    return tensor.detach().cpu().numpy()

@svc.api(input=Image(), output=NumpyNdarray(dtype="int64"))
async def predict_image(f: PILImage) -> NDArray[t.Any]:
    arr = np.expand_dims(arr, (0, 1)).astype("float32")
    
    # 이런식으로 분산 처리 하는게 가능할것으로 보인다.
    if xxx: # random 1/2
         mnist_runner1.async_run(arr)
    else:
        mnist_runner2.async_run(arr)
    

    runner1_result, runner2_result = results  # [tensor([5]), tensor([5])]

    bentoml_logger.info(f"111111111: {to_numpy(runner1_result)[0]}")
    bentoml_logger.info(f"2222222: {to_numpy(runner2_result)[0]}")

    return to_numpy(runner1_result)

~~~

---

이 코드는 Multi GPU 환경에서 Runner를 어떻게 2개를 띄울수 있는지 에 대한 문의와
비슷한 것 같아 작년 12월쯔음 discussion 에 올라왔던 답변 첨부한다 
https://github.com/bentoml/BentoML/discussions/3303#discussioncomment-5695387
